README.md
# J.A.M.E.S. 

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar james-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-05-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-05-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/james/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/james/compare/0.1.0...0.1.1


test/james/core_test.clj
(ns james.core-test
  (:require [clojure.test :refer :all]
            [james.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/james/core/tools.clj
(ns james.core.tools)




(defn uuid
  "simple UUID creator"
  [] (str (java.util.UUID/randomUUID)))

(defn flatten2
  "Like `clojure.core/flatten` but better, stronger, faster.
  Takes any nested combination of sequential things (lists, vectors,
  etc.) and returns their contents as a single, flat, lazy sequence.
  If the argument is non-sequential (numbers, maps, strings, nil,
  etc.), returns the original argument."
  {:static true}
  [x]
  (letfn [(flat [coll]
                  (lazy-seq
                   (when-let [c (seq coll)]
                     (let [x (first c)]
                       (if (sequential? x)
                         (concat (flat x) (flat (rest c)))
                         (cons x (flat (rest c))))))))]
    (if (sequential? x) (flat x) x)))

(defn put-key-prefix
  [attrs prefix]
  (let [ keys-of-atts (keys attrs)
         replacement (into {} (map (fn[k] [k (->> k name (str (name prefix) "-") keyword)]) keys-of-atts ))]
    (clojure.set/rename-keys attrs replacement)))



; ################ TIME STAFF

(require '[clj-time.core :as time]
         '[clj-time.coerce :as tc])

(defn now
  ""
  [] (tc/to-long (time/now)))

(defn month-by-ts
  [ts]
  (time/month (tc/from-long ts)))

(defn day-by-ts
  [ts]
  (time/day (tc/from-long ts)))

(defn weekday-by-ts
  [ts]
  (time/day-of-week (tc/from-long ts)))

(defn hour-by-ts
  [ts]
  (time/hour (tc/from-long ts)))

(defn decorate-time
  [m]
  (let [date-ts (:date m)]
    (merge m {:day-of-week (weekday-by-ts date-ts), :day-of-month (day-by-ts date-ts), :month  (month-by-ts date-ts), :hour (hour-by-ts date-ts)})))




src/james/core/decision/tree.clj
(ns james.core.decision.tree
  (:use james.core.datamodel.nodes)
  (:use james.core.tools))


(defn log2 [x]
  "Return the base-2 logarithm of x"
  (/ (Math/log x) (Math/log 2)))


(defn entropy [alist]
  "Given a list of pairs if (input, output), calculate the entropy of the list with respect to output"
  (let [n (count alist)
        counts (vals (frequencies (map #(nth % 1) alist)))]
    (reduce + (map #(* (- %) (log2 %)) (map #(/ % n) counts)))))


(defn partition-map
  "Given function and a seq, return a map of values the function can take when using members of the seq as an argument, to sub-seqs containing the parts of seq corresponding to those outputs"
  ([afn aseq]
   (partition-map {} afn aseq))
  ([acc afn aseq]
   (if (empty? aseq)
     acc
     (let [i (first aseq) o (afn i) inputs (get acc o)]
       (if inputs
         (recur
           (assoc acc o (conj inputs i)) afn (rest aseq))
         (recur
           (assoc acc o [i]) afn (rest aseq)))))))



(defn all-keys [aseq]
  "Given a list of maps, return a set containing all the keys from those maps"
  (into #{} (flatten2 (map keys aseq))))


(defn gain
  ([k alist]
   "Given a key and a list of pairs of (map, output), return the net reduction in entropy caused by partitioning the list according to the values associated with key in the input maps"
   (gain (entropy alist) k alist))
  ([current-entropy k alist]
   (let [parts (vals (partition-map #(k (first %)) alist))]
     (- current-entropy
        (reduce +
                (map #(* (entropy %) (/ (count %) (count alist))) parts))))))


(defn most-informative-key [aseq]
  "Given a sequence of pairs of (map, output), return the map key that provides the highest information gain about outout, when used to partition the sequence"
  (apply max-key #(gain % aseq) (all-keys (map first aseq))))


(defn map-vals
  ([amap fn]
   "Apply fn to each of the values in amap, and return a map associating those keys with the new values"
   (map-vals {} amap fn))
  ([acc amap fn]
   (if (empty? amap)
     acc
     (let [[k v] (first amap)]
       (recur (assoc acc k (fn v)) (rest amap) fn)))))

(defn build-decision-tree [aseq]
  "A decision tree is either: a pair consisting an input-map key, followed by a map of values to further trees, or a leaf node: a result (an output, in the language of the rest of this library). aseq should be a sequence of pairs of [input-map, output]. Returns nil if the seq is empty."
  (if (zero? (entropy aseq))
    (nth (first aseq) 1)
    (let [k (most-informative-key aseq)]
      [k (map-vals (partition-map #(k (first %)) aseq) build-decision-tree)])))


(defn tree-decide [tree input-map]
  "Given a decision tree and an input map, make decisions based on the contents of input-map and return the correct output (leaf node)."
  (if (vector? tree)
    (let [[k parts-map] tree]
      (tree-decide (parts-map (k input-map)) input-map))
    tree))


(defn comb [k l]
  (if (= 1 k) (map vector l)
      (apply concat
             (map-indexed
              #(map (fn [x] (conj x %2))
                    (comb (dec k) (drop (inc %1) l)))
              l))))


(defn all-subsets [s]
  (apply concat
         (for [x (range 1 (inc (count s)))]
           (map #(into #{} %) (comb x s)))))

; (all-subsets #{1 2 3}) ;-> (#{1} #{3} #{2} #{1 3} #{1 2} #{3 2} #{1 3 2})

(defn all-submaps [m]
  (apply concat
         (for [x (range 1 (inc (count m)))]
           (map #(into {} %) (comb x m)))))

;(all-submaps {:a 1, :b 2, :c 3}) ; -> ({:a 1} {:b 2} {:c 3} {:b 2, :a 1} {:c 3, :a 1} {:c 3, :b 2} {:c 3, :b 2, :a 1})


(defn most-frequent-n [n items]
  (->> items
    frequencies
    (sort-by val)
    reverse
    (take n)
    (map first)))

;(most-frequent-n 2 [1 2 4 56 7 56 3 2 2 4 56 7 1 54 45 6 73 3 4 34 2 2 56]) ; -> (2 56)



(defn get-decision-for
  ([pname cat?]
  (let [tree (build-decision-tree (get-decision-data cat?))
        timemap (put-key-prefix (decorate-time {:date (now)}) :ordered)
        feed (merge timemap (get-last-order pname))]
    (println feed)
    (first (most-frequent-n 1 (filter #(not (nil? %))
                                      (map #(tree-decide tree %) (all-submaps feed)))))))
  ([pname]
   (get-decision-for pname true)))


; #####################################################################################################

(build-decision-tree (get-decision-data true))

(get-decision-for "Sören")
(get-decision-for "Silvia")
(get-decision-for "Heiko")







resources/defaultgraphdata/data.clj
(add-order-with-rating "1"
                       (add-food "Pizza-Brötchen"
                                 (find-or-create-deliverer "Napoli" {}) {})
                       1
                       1464192331000 5 "")

(add-order-with-rating "1"
                       (add-food "Mam Burger"
                                 (find-or-create-deliverer "Napoli" {}) {})
                       1
                       1464192331000 3 "")

(add-order-with-rating (find-or-create-person "Heiko" {})
                       (add-food "Pizza-Brötchen"
                                 (find-or-create-deliverer "Napoli" {}) {})
                       1
                       1464192331000 4 "")


(add-order-with-rating (find-or-create-person "Heiko" {})
                       (add-food "Pizza"
                                 (find-or-create-deliverer "Napoli" {}) {})
                       1
                       1464192331000 4 "")

(add-order-with-rating "1"
                       (add-food "Ente süß-sauer"
                                 (find-or-create-deliverer "China Boy" {}) {})
                       1
                       1464283100347 2 "")

(add-order-with-rating "1"
                       (add-food "Pizza-Brötchen"
                                 (find-or-create-deliverer "Napoli" {}) {})
                       1
                       1458492283000 4 "")

(add-order-with-rating "1"
                       (add-food "Pizza"
                                 (find-or-create-deliverer "Napoli" {}) {})
                       1
                       1458492283000 3 "")


src/james/core/datamodel/nodes.clj
(ns james.core.datamodel.nodes
  (:use james.core.tools)
  (:use loom.graph loom.attr loom.io loom.derived loom.label loom.alg)
  (:use clojure.java.io)
  (:require [clojure.edn    :as edn]))


(def vertex-types #{:default, :person, :order, :deliverer, :rating, :picture, :order-pos, :food})

(def edge-types #{:created :default, :ordered, :comesFrom, :rates, :lookslike, :hasRating})

(def file-path "/opt/james/brain.graph")


(defn date-properties
  [date-ts]
  {:date date-ts})


(defn new-vertex
  [g id vertex-name vertex-type properties]
  (let [ id (if (= id "")
              (uuid)
              id)
         vertex-type (if (contains? vertex-types vertex-type)
                       vertex-type
                       :default)
         properties (if (map? properties)
                      properties
                      (if (empty? properties)
                        {}
                        {:default properties}))
         g (add-nodes g id)]
    (reduce (fn [g [k v]] (add-attr g id k v)) g (merge (date-properties (now))
                                                        {:label (str vertex-name " " id), :name vertex-name, :type vertex-type}
                                                        properties))))


(defn new-edge
  ([g edge-type source target ]
   (new-edge g edge-type source target {}))
  ([g edge-type source target properties]
   (let [ edge-type (if (contains? edge-types edge-type)
                      edge-type
                      :default)
          properties (if (map? properties)
                       properties
                       (if (empty? properties)
                         {}
                         {:default properties}))
          g (add-edges g [source target])]
     (reduce (fn [g [k v]] (add-attr-to-edges g k v [[source target]] )) g (merge (date-properties (now))
                                                                                  {:label edge-type
                                                                                   :type edge-type}
                                                                                  properties)))))



(defn find-vertices
  [g attr]
  (let [graph-attrs (:attrs g)
        ks (keys attr)]
    (keys (filter (fn [[k v]] (= attr (select-keys v ks))) graph-attrs))))


(defn find-edge
  [g n1 n2 attr]
  )

(defn save-memory
  [graph]
  (spit file-path  (prn-str graph)))


(defn create-default-graph
  []
  (-> (graph)
      (new-vertex "1" "Sören" :person {:role "admin" :birth "13-09-1989"})))


(defn load-existing-memory
  []
  (if (.exists (as-file file-path))
    (edn/read-string {:readers {'loom.graph.BasicEditableGraph loom.graph/map->BasicEditableGraph}} (slurp file-path))
    (create-default-graph)))


(def james-graph (ref (load-existing-memory)))


(defn add-person
  [name ops creator-id?]
  (let [creator-id (if (nil? creator-id?)
                     "1"
                     creator-id?)
        id (uuid)]
    (dosync
      (ref-set james-graph  (new-vertex @james-graph id name :person ops))
      (save-memory @james-graph)
      id)))


(defn add-deliverer
  [name ops]
  (let [id (uuid)]
    (dosync
      (ref-set james-graph (new-vertex @james-graph id name :deliverer ops))
      (save-memory @james-graph))
    id))



(defn add-food
  [foodname deliverer-id ops ]
  (let [id (uuid)]
    (dosync
      (ref-set james-graph (new-edge (new-vertex @james-graph id foodname :food ops)
                                     :comesFrom id deliverer-id {}))
      (save-memory @james-graph))
    id))


(defn add-order
  [person-id food-id n date?]
  (let [date (if (nil? date?)
               (now)
               date?)]
    (dosync
      (ref-set james-graph (new-edge @james-graph :ordered person-id food-id (merge {:count n} (date-properties date))))
      (save-memory @james-graph))
    [person-id food-id]))



(defn add-rating
  [person-id food-id rating ops]
  (let [id (uuid)]
    (dosync
      (ref-set james-graph (new-edge (new-edge (new-vertex @james-graph id (str rating "*") :rating ops)
                                               :hasRating food-id id {})
                                     :rates person-id id {}))
      (save-memory @james-graph)
      id)))


(defn add-order-with-rating
  [person-id food-id n date? rating note?]
  (let [ordering (add-order person-id food-id n date?)
        ]
    (add-rating person-id food-id rating {:note note?})))


(defn find-vertex
  [g name type]
  (let [results (find-vertices g {:type type, :name name})]
    (if (not (empty? results))
      (first results)
      nil)))



(defn find-or-create-person
  [name ops]
  (let [result (find-vertex @james-graph name :person)]
    (if (nil? result)
      (add-person name ops nil)
      result)))


(defn find-or-create-deliverer
  [name ops]
  (let [result (find-vertex @james-graph  name :deliverer)]
    (if (nil? result)
      (add-deliverer name ops)
      result)))


(defn- get-edges-by-type
  [g node type]
  (sequence  (comp
               (filter (fn[edge] (contains? (set edge) node)))
               (filter (fn[edge] (= type (attr @james-graph (first edge) (second edge) :type)))))
             (distinct-edges (subgraph-reachable-from @james-graph node))))




(defn- distinct-by
  "Returns a lazy sequence of the elements of coll, removing any elements that
  return duplicate values when passed to a function f."
  ([f]
   (fn [rf]
     (let [seen (volatile! #{})]
       (fn
         ([] (rf))
         ([result] (rf result))
         ([result x]
          (let [fx (f x)]
            (if (contains? @seen fx)
              result
              (do (vswap! seen conj fx)
                (rf result x)))))))))
  ([f coll]
   (let [step (fn step [xs seen]
                (lazy-seq
                  ((fn [[x :as xs] seen]
                     (when-let [s (seq xs)]
                       (let [fx (f x)]
                         (if (contains? seen fx)
                           (recur (rest s) seen)
                           (cons x (step (rest s) (conj seen fx)))))))
                   xs seen)))]
     (step coll #{}))))






(defn get-last-order
  [pname]
  (let [g @james-graph
        pid (first (find-vertices g {:type :person :name pname}))
        orders (map (fn [[_ oid]] [(attrs g  oid)
                                   (attrs g pid oid)
                                   (attrs g (last (last (get-edges-by-type g oid :comesFrom))))
                                   (attrs g (last (last (get-edges-by-type g oid :hasRating))))]) (get-edges-by-type g pid :ordered))
       [lastfood lastorder lastdeliverer lastrating] (apply max-key #(:date (second %)) orders)]
  {:person-name pname
   :further-food-cat (:category lastfood)
   :further-deliverer (:name lastdeliverer)
   :further-food (:name lastfood)
   :further-rating (:name lastrating)}))



(defn travel
  [nodes edge-type]
  (let [nodes (if (coll? nodes)
                (if (coll? (first nodes))
                  (into #{} (map set nodes))
                  #{(set nodes)})
                #{#{nodes}})
        g @james-graph ]
    (into #{} (flatten (map (fn [ns]  (filter #(and (not (nil? %)) (coll? %) (not (empty? %)))
                                              (map (fn[n] (when-not (vector? n)
                                                            (map #(apply conj ns % %) (get-edges-by-type g n edge-type)))) ns))) nodes)))))





(defn value-maps-by-ids
  [ids]
  (map (fn [e] (reduce (fn[m id] (let [attrs (if (vector? id)
                                               (attrs @james-graph (first id) (second id))
                                               (attrs @james-graph id))]
                                   (assoc m (:type attrs) attrs))) {} e)) ids))



(defn get-decision-data
  [cat?]
  (let [g @james-graph
        people (find-vertices g {:type :person})
        people-and-orders  (flatten (map #(-> %
                                              (travel :ordered)
                                              (travel :comesFrom)
                                              (travel :hasRating)
                                              (value-maps-by-ids)) people))
        get-prefixed-attrs (fn[m prefix nogos]
                             (put-key-prefix
                               (apply dissoc m :label :type nogos)
                               prefix))
        raw-orders (map (fn [maps] (merge (get-prefixed-attrs (:person maps) :person #{:date})
                                          (get-prefixed-attrs (:food maps) :food #{:date})
                                          (get-prefixed-attrs (:rating maps) :rating #{:date :note})
                                          (get-prefixed-attrs (:deliverer maps) :deliverer #{:date})
                                          (get-prefixed-attrs (decorate-time (:ordered maps)) :ordered #{})))
                        people-and-orders)
        orders-by-peoples (partition-by :person-name raw-orders)]
    (into [] (apply concat (map (fn [list-of-orders] (let [sorted-list (sort-by :ordered-date  (distinct-by :ordered-date  list-of-orders))]
                                                       (mapv (fn[a b] [ (dissoc
                                                                          (assoc a
                                                                            :further-food (:food-name b)
                                                                            :further-food-cat (:food-categorie b)
                                                                            :further-rating (:rating-name b)
                                                                            :further-deliverer (:deliverer-name b)) :food-category :person-birth :person-role :food-name :rating-name :deliverer-name :ordered-date :ordered-count )
                                                                        (select-keys a (if cat? [:food-category :deliverer-name] [:food-name :deliverer-name]))])
                                                             sorted-list (conj sorted-list {})))) orders-by-peoples)))))





; ############################################

;(def traindata (get-decision-data false))



; Some default data

#_(
    (add-order-with-rating "1"
                          (add-food "Schnitzel Napoli"
                                    (find-or-create-deliverer "Schnitzelhaus" {}) {:category "Schnitzel"})
                          1
                          1459182354000 3 "")

    (add-order-with-rating (find-or-create-person "Silvia" {})
                          (add-food "Chicken Wrap"
                                    (find-or-create-deliverer "Schnitzelhaus" {}) {:category "Wrap"})
                          1
                          1459182354000 4 "")

    (add-order-with-rating "1"
                          (add-food "Drehspieß Wrap"
                                    (find-or-create-deliverer "Schnitzelhaus" {}) {:category "Wrap"})
                          1
                          1459182354000 2 "")

    (add-order-with-rating "1"
                          (add-food "Gyros Teller"
                                    (find-or-create-deliverer "Vulkan" {}) {:category "Griechisch"})
                          1
                          1460929074000 3 "")

    (add-order-with-rating "1"
                          (add-food "Pizza Tonno"
                                    (find-or-create-deliverer "Himalaya" {}) {:category "Pizza"})
                          1
                          1459282374000 4 "")

      (add-order-with-rating "1"
                          (add-food "Pizza Tonno"
                                    (find-or-create-deliverer "Himalaya" {}) {:category "Pizza"})
                          1
                          1461510294000 4 "")

   (add-order-with-rating (find-or-create-person "Silvia" {})
                          (add-food "Pizza Quattro Stagioni"
                                    (find-or-create-deliverer "Himalaya" {}) {:category "Pizza"})
                          1
                          1461510294000 4 "")

    (add-order-with-rating "1"
                          (add-food "Hawaii-Schnitzel"
                                    (find-or-create-deliverer "Schnitzelhaus" {}) {:category "Schnitzel"})
                          1
                          1462478874000 4 "")

    (add-order-with-rating (find-or-create-person "Silvia" {})
                          (add-food "Pasta Classico"
                                    (find-or-create-deliverer "Schnitzelhaus" {}) {:category "Pasta"})
                          1
                          1462478874000 2 "")

   (add-order-with-rating "1"
                          (add-food "Pizza-Brötchen"
                                    (find-or-create-deliverer "Napoli" {}) {:category "Beilage"})
                          1
                          1464192331000 5 "")

   (add-order-with-rating "1"
                          (add-food "Mam Burger"
                                    (find-or-create-deliverer "Napoli" {}) {:category "Burger"})
                          1
                          1464192331000 3 "")

   (add-order-with-rating (find-or-create-person "Heiko" {})
                          (add-food "Pizza-Brötchen"
                                    (find-or-create-deliverer "Napoli" {}) {:category "Beilage"})
                          1
                          1464192331000 4 "")


   (add-order-with-rating (find-or-create-person "Heiko" {})
                          (add-food "Pizza"
                                    (find-or-create-deliverer "Napoli" {}) {:category "Pizza"})
                          1
                          1464192331000 4 "")

   (add-order-with-rating "1"
                          (add-food "Ente süß-sauer"
                                    (find-or-create-deliverer "China Boy" {}) {:category "Chinesisch"})
                          1
                          1464283100347 2 "")

   (add-order-with-rating "1"
                          (add-food "Pizza-Brötchen"
                                    (find-or-create-deliverer "Napoli" {}) {:category "Beilage"})
                          1
                          1458492283000 4 "")

   (add-order-with-rating "1"
                          (add-food "Pizza Momento"
                                    (find-or-create-deliverer "Napoli" {}) {:category "Pizza"})
                          1
                          1458492283000 3 "")

   (add-order-with-rating "1"
                          (add-food "Pizza Hawaii"
                                    (find-or-create-deliverer "Himalaya" {}) {:category "Pizza"})
                          1
                          1464552710045 3 "")

   (add-order-with-rating (find-or-create-person "Silvia" {})
                          (add-food "Pizza Tonno"
                                    (find-or-create-deliverer "Himalaya" {}) {:category "Pizza"})
                          1
                          1464552710045 3 "")

  (add-order-with-rating (find-or-create-person "Heiko" {})
                          (add-food "Enchilada"
                                    (find-or-create-deliverer "Vulkan" {}) {:category "Wrap"})
                          1
                          1465334571000 5 "")
   )



;(nodes-filtered-by #(= "Sören" (:name %)) data-graph)
;(view @james-graph)
